scripts/inject-relup.escript

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env escript
+
+%% This script injects implicit relup instructions for emqx applications.
+
+-mode(compile).
+
+-define(ERROR(FORMAT, ARGS), io:format(standard_error, "[inject-relup] " ++ FORMAT ++ "~n", ARGS)).
+-define(INFO(FORMAT, ARGS), io:format(user, "[inject-relup] " ++ FORMAT ++ "~n", ARGS)).
+
+usage() ->
+    "Usage: " ++ escript:script_name() ++ " <path-to-release-dir> <release-vsn>".
+
+main([RelRootDir, CurrRelVsn]) ->
+    case filelib:is_dir(filename:join([RelRootDir, "releases"])) andalso
+         filelib:is_dir(filename:join([RelRootDir, "lib"])) of
+        true ->
+            EmqxAppVsns = get_emqx_app_vsns(RelRootDir),
+            ok = inject_relup_file(RelRootDir, CurrRelVsn, EmqxAppVsns);
+        false ->
+            ?ERROR("not a valid root dir of release: ~p, for example: _build/emqx/rel/emqx",
+                [RelRootDir]),
+            erlang:halt(1)
+    end;
+main(_Args) ->
+    ?ERROR("~s", [usage()]),
+    erlang:halt(1).
```

Comment:
do it like this:
```
stop() ->
    init:stop(1),
    timer:sleep(100),
    stop().
```
otherwise the io:format prints may race with halt.

Reply:
If the first `init:stop(1)` is running but not finished, will it work to try again?

Reply:
And I understand the `io:format/2` is an sync operation, is there really a race condition?

Reply:
It happened to cuttlefish.
And it’s solved by doing the recursive stop.
https://github.com/emqx/cuttlefish/blob/6c346563e89ebbd95dbc1c29017adaf9abf85ca1/src/cuttlefish_escript.erl#L251

Reply:
That's probably it uses `logger` and `cuttlefish_test_logger`, which are async (by default).

I did the following experiment, it shows `io:format/2` always prints the msg before the vm is halt, but the logger can't:

```
{22-03-01 13:43}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> logger:error("hahah"), halt(1).

{22-03-01 13:43}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> logger:error("hahah"), halt(1).

{22-03-01 13:43}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> io:format("hahah"), halt(1).
hahah%

{22-03-01 13:44}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> io:format("hahah"), halt(1).
hahah%
```